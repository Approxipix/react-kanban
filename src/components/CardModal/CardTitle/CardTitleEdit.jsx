import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { editCardTitle } from '../../../redux/cardReducer/actions';
import ClickOutside from '../../ClickOutside';
import { Input, TextArea, } from '../../BaseComponent';
import styled from 'styled-components'

const Form = styled.form`
  width: calc(100% - 2rem);
`;

class CardTitleEdit extends Component {
  constructor(props) {
    super(props);
    this.state = {
      title: props.cardTitle || '',
    };
  }

  handleChange = (value) => {
    this.setState({
      title: value,
    });
  };

  handleKeyDown = (e) => {
    if (e.keyCode === 27) {
      this.props.toggleEditTitle();
    }
  };

  handleSubmit = (e) => {
    e.preventDefault();
    const { title } = this.state;
    const { cardId } = this.props;
    if (!title)  return;
    this.props.actions.editCardTitle({
      cardTitle: title,
      cardId: cardId,
    });
    this.props.toggleEditTitle()
  };

  render() {
    const { title } = this.state;
    return (
      <ClickOutside toggleOpened={this.props.toggleEditTitle}>
        <Form onSubmit={this.handleSubmit}>
          <Input
            type="text"
            margin="0"
            padding=".3rem .5rem"
            size="1.4rem"
            weight="600"
            value={title}
            placeholder="Edit card title"
            onKeyDown={this.handleKeyDown}
            onChange={(e) => this.handleChange(e.target.value)}
            onBlur={this.handleSubmit}
            spellCheck={false}
            autoFocus
          />
        </Form>
      </ClickOutside>
    );
  };
}

function mapDispatchToProps(dispatch) {
  return {
    actions: bindActionCreators({
      editCardTitle: editCardTitle,
    }, dispatch)
  };
}

export default connect(null, mapDispatchToProps)(CardTitleEdit);
