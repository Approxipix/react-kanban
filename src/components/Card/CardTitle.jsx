import React, { Component } from 'react';
import styled from 'styled-components';
import CardTitleEdit from './CardTitleEdit';

const Title = styled.h4`
  margin-bottom: .5rem;
  padding: .3rem .5rem;
  font-size: 1.5rem;
  font-weight: 600;
`;

class CardTitle extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isTitleEditing: false,
    }
  }

  toggleIsTitleEditing = () => {
    this.setState({
      isTitleEditing: !this.state.isTitleEditing
    })
  };

  render() {
    const { cardId, cardTitle } = this.props;
    const { isTitleEditing } = this.state;

    if (isTitleEditing) {
      return (
        <CardTitleEdit
          cardId={cardId}
          cardTitle={cardTitle}
          toggleIsTitleEditing={this.toggleIsTitleEditing}
        />
      )
    }
    return (
      <Title onClick={() => this.toggleIsTitleEditing()}>
        {cardTitle}
      </Title>
    )
  }
}


export default CardTitle;
