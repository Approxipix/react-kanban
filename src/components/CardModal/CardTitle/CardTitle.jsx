import React, { Component } from 'react';
import PropTypes from "prop-types";
import CardTitleEdit from './CardTitleEdit';
import styled from 'styled-components';

const Title = styled.h2`
  margin-bottom: .5rem;
  padding: .3rem .5rem;
  font-size: 1.4rem;
  font-weight: 600;
  white-space: nowrap;
  text-overflow: ellipsis;
  overflow: hidden;
`;
Title.displayName = 'Title';

class CardTitle extends Component {
  constructor(props) {
    super(props);
    this.state = {
      editTitle: false,
    }
  }

  toggleEditTitle = () => {
    this.setState({
      editTitle: !this.state.editTitle
    })
  };

  render() {
    const { cardId, cardTitle } = this.props;
    const { editTitle } = this.state;

    return (
      !editTitle ? (
        <Title onClick={() => this.toggleEditTitle()}>
          {cardTitle}
        </Title>
      ) : (
        <CardTitleEdit
          cardId={cardId}
          cardTitle={cardTitle}
          toggleEditTitle={this.toggleEditTitle}
        />
      )
    )
  }
}

CardTitle.defaultProps = {
  cardId: '',
  cardTitle: '',
};

CardTitle.propTypes = {
  cardId: PropTypes.string.isRequired,
  cardTitle: PropTypes.string.isRequired,
};

export default CardTitle;
