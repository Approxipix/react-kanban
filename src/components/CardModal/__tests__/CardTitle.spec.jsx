import React from 'react';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import CardTitle from '../CardTitle/CardTitle';

describe('<CardTitle>', () => {
  let state;
  let component;

  beforeEach(() => {
    state = {
      editTitle: false,
    };

    component = shallow(<CardTitle />);
    component.setState({...state});
  });

  it('should render title if editTitle is false', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render CardTitleEdit component if editTitle is true', () => {
    component.instance().toggleEditTitle();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('checks if toggleEditTitle sets editTitle correctly', () => {
    component.instance().toggleEditTitle();
    expect(component.instance().state.editTitle).toEqual(true);

    component.instance().toggleEditTitle();
    expect(component.instance().state.editTitle).toEqual(false);
  });

  it("should call toggleEditTitle after click on title", () => {
    const spy = jest.spyOn(component.instance(), "toggleEditTitle");

    component.find('Title').simulate('click');

    expect(spy).toHaveBeenCalledTimes(1);
  });
});
