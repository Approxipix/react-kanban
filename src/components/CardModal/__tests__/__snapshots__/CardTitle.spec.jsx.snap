// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardTitle> should render CardTitleEdit component if editTitle is true 1`] = `
<Connect(CardTitleEdit)
  cardId=""
  cardTitle=""
  toggleEditTitle={[Function]}
/>
`;

exports[`<CardTitle> should render title if editTitle is false 1`] = `
<Title
  onClick={[Function]}
/>
`;
